drivers: pwm: add RP2040 PWM support

burumaj opened the pull request:
Add PWM driver support for the RP2040 using pinctrl

yonsch commented on drivers/pwm/pwm_rpi_pico_slice.c:
```diff
@@ -0,0 +1,147 @@
+{
+	struct pwm_rpi_data *data = dev->data;
+	const struct pwm_rpi_config *cfg = dev->config;
+	bool pwm_polarity_a = (data->slice->csr & PWM_CH0_CSR_A_INV_BITS) > 0;
```
The bool to bit conversion (using >) isn't necessary, as `pwm_set_output_polarity` converts it using `bool_to_bit`

burumaj replied:
This is not a conversion from bool to but, but from bit to bool. This is then used to later invert the selected channel.
Then later on `pwm_set_output_polarity` calls `bool_to_bit` to convert it to a bit again.

yonsch commented on drivers/pwm/pwm_rpi_pico_slice.c:
```diff
@@ -0,0 +1,147 @@
+	}
+
+	pwm_rpi_set_channel_polarity(dev, ch,
+				     (flags && PWM_POLARITY_MASK) == PWM_POLARITY_INVERTED);
```
Is the `&&` instead of `&` intentional? This expression will evaluate to true if any flag is set

burumaj replied:
fixed

yonsch reviewed (commented):
My initial review, for what it's worth. I will also try to test this myself

burumaj commented on modules/hal_rpi_pico/Kconfig:
```diff
@@ -13,3 +13,8 @@ config PICOSDK_USE_GPIO
+config PICOSDK_USE_PWM
+	bool
+	help
+	  Use the PWM driver form pico-sdk
```
fixed

burumaj commented on boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi:
```diff
@@ -15,4 +15,10 @@
+
+	pwm_ch4b_default: pwm_ch4b_default {
+		group1 {
+			pinmux = < PWM_4B_P25 >;
```
fixed

burumaj commented on boards/arm/rpi_pico/rpi_pico.dts:
```diff
@@ -26,8 +27,18 @@
+	pwm_leds {
+		compatible = "pwm-leds";
+		pwm_led0: pwm_led_0 {
+			status = "okay";
```
No it's not. Fixed.

burumaj commented on boards/arm/rpi_pico/rpi_pico.dts:
```diff
@@ -68,3 +79,14 @@
+};
+
+&pwm4 {
+	status = "okay";
```
Fixed

burumaj commented on drivers/pwm/Kconfig.rpi_pico_slice:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config PWM_RPI_PICO
+    bool "RP2040 PWM"
```
fixed

burumaj commented on drivers/pwm/Kconfig.rpi_pico_slice:
```diff
@@ -0,0 +1,10 @@
+config PWM_RPI_PICO
+    bool "RP2040 PWM"
+		default y
+		depends on SOC_SERIES_RP2XXX
```
should be fixed now

burumaj commented on drivers/pwm/pwm_rpi_pico_slice.c:
```diff
@@ -0,0 +1,147 @@
+
+struct pwm_rpi_config {
+	/* pwm_controller is the start address of the pwm peripheral.
+	 * This address is used to get a refference to pwm_slice_hw,
```
fixed

burumaj commented on drivers/pwm/CMakeLists.txt:
```diff
@@ -24,6 +24,7 @@ zephyr_library_sources_ifdef(CONFIG_PWM_XLNX_AXI_TIMER	pwm_xlnx_axi_timer.c)
 zephyr_library_sources_ifdef(CONFIG_PWM_MCUX_PWT 	pwm_mcux_pwt.c)
 zephyr_library_sources_ifdef(CONFIG_PWM_GECKO   pwm_gecko.c)
 zephyr_library_sources_ifdef(CONFIG_PWM_GD32 pwm_gd32.c)
+zephyr_library_sources_ifdef(CONFIG_PWM_RPI_PICO	pwm_rpi_pico_slice.c)
```
This can be `pwm_rpi_pico`. This has mostly got to do with the way the drivers is implemented right now.
It exposes a device for each individual pwm slice. I had some confusion regarding the pwm api. I'll rewrite
the driver to expose a single pwm device with 16 channels. I'll try to update this before the weekend.

burumaj replied:
With the latest push this should be fixed

burumaj commented:
I'll clean up the commits after a bit of rework of the driver.

burumaj committed:
dts: arm: rpi_pico: add dt bindings for pwm

Add devicetree bindings for pwm peripheral using pinctrl

Signed-off-by: Joep Buruma <burumaj50@gmail.com>

burumaj committed:
drivers: pwm: add pwm driver for rpi_pico

Add pwm driver implementation for rpi pico family.
Adds myself as codeowner of drivers/pwm/*rpi_pico*

Signed-off-by: Joep Buruma <burumaj50@gmail.com>

burumaj committed:
boards: arm: rpi_pico: add pwm bindings to devictree

Add bindings for pwm-leds to the rpi pico board. The default
pwm binding is connected to the onboard LED (P25).

Signed-off-by: Joep Buruma <burumaj50@gmail.com>

burumaj committed:
samples: led_pwm: add overlay for rpi_pico board

Add an overlay for rpi_pico board to the led_pwm sample.
This board only supports fairly high pwm frequencies so
the blinking of the LED is not supported. Turning it on/off
and fading is.

Signed-off-by: Joep Buruma <burumaj50@gmail.com>

gmarull commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -111,6 +111,17 @@
+		
+		pwm: pwm@40050000{
+			compatible = "raspberrypi,pico-pwm";
+			reg = < 0x40050000 DT_SIZE_K(4) >;
```
`reg = <0x40050000 DT_SIZE_K(4)>;`

gmarull commented on drivers/pwm/Kconfig.rpi_pico:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config PWM_RPI_PICO
+	bool "RPi Pico PWM"
+		default y
+		depends on SOC_SERIES_RP2XXX
+	select PICOSDK_USE_PWM
```
please fix formatting here, and default based on dt_compat_status (check some other drivers)

gmarull commented on modules/hal_rpi_pico/CMakeLists.txt:
```diff
@@ -78,4 +78,8 @@ if(CONFIG_HAS_RPI_PICO)
   zephyr_include_directories_ifdef(CONFIG_PICOSDK_USE_UART
           ${rp2_common_dir}/hardware_uart/include)
+
+  zephyr_include_directories_ifdef(CONFIG_PICOSDK_USE_PWM
+          ${rp2_common_dir}/hardware_pwm/include)
```
question: is it a header only driver?

burumaj replied:
Yes, this is header only driver

gmarull commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -111,7 +111,7 @@
 			label = "I2C_1";
 			status = "disabled";
 		};
-		
```
nit: seems like an artifact? were tabs initially here?

gmarull commented on samples/drivers/led_pwm/boards/rpi_pico.overlay:
```diff
@@ -0,0 +1,7 @@
+&pwm_led0 {
+	status = "okay";
+};
+
+&pwm {
+	status = "okay";
+};
```
can you please explain this better? Why can't the sample run on rpi pico?

burumaj replied:
The PWM peripheral is clocked by `&system_clk` (will fix this in the driver), which by default runs at 125MHz on the `rpi_pico` board. This can be divided down by ~256 (255 15/16 in fact) resulting in a frequency of ~490 KHz as the lowest frequency that can be achieved by the pwm peripheral, resulting in a period of ~2 us. The pwm_led example tries to blink the onboard LED using the pwm peripheral with a period of 0.1 s. The part of the sample that does work is fading the LED since this requires far shorter periods of the PWM peripheral.

gmarull commented on drivers/pwm/pwm_rpi_pico.c:
```diff
@@ -0,0 +1,178 @@
+		(float)cfg->slice_configs[slice].integral / 16.0;
+}
+
+static inline int pwm_rpi_channel_to_slice(uint32_t channel)
+{
+	return channel / 2;
+}
+
+static inline int pwm_rpi_channel_to_pico_channel(uint32_t channel)
+{
+	return channel % 2;
+}
```
nit seems like you are returning `int` from `uint32_t`, should likely return `uint32_t`, or, cast down if you're sure about the range of values.

gmarull commented on drivers/pwm/pwm_rpi_pico.c:
```diff
@@ -0,0 +1,178 @@
+		slice_cfg = pwm_get_default_config();
+		pwm_config_set_clkdiv_mode(&slice_cfg, PWM_DIV_FREE_RUNNING);
+
+		pwm_init(i, &slice_cfg, false);
+
+		pwm_set_clkdiv_int_frac(i,
+					cfg->slice_configs[i].integral,
+					cfg->slice_configs[i].frac);
+		pwm_set_enabled(i, true);
```
a general comment, not related to this PR: I find RPi HAL really confusing since it uses the same Zephyr prefix (happens in other areas as well), I'd personally use register access, I think it would lead to more readable drivers + less overhead.

henrikbrixandersen commented:
@burumaj Please rebase.

yonsch commented on modules/hal_rpi_pico/CMakeLists.txt:
```diff
@@ -99,4 +102,5 @@ if(CONFIG_HAS_RPI_PICO)
     PROPERTIES
     COMPILE_FLAGS $<TARGET_PROPERTY:compiler,warning_no_pointer_arithmetic>
   )
+
```
remove

burumaj replied:
done

yonsch reviewed (changes requested):
Looking good, added a few comments. Also, make PWM a reset device. See pull requests for other drivers for reference.

burumaj commented on dts/bindings/pwm/raspberrypi,pico-pwm.yaml:
```diff
@@ -0,0 +1,129 @@
+
+compatible: "raspberrypi,pico-pwm"
+
+include: [pwm-controller.yaml, pinctrl-device.yaml, base.yaml]
```
Apparently, yes. All included bindings never include `base.yaml`

burumaj commented on boards/arm/rpi_pico/rpi_pico.dts:
```diff
@@ -28,8 +29,18 @@
+		compatible = "pwm-leds";
+		pwm_led0: pwm_led_0 {
+			status = "disabled";
+			pwms = < &pwm 9 500 PWM_POLARITY_NORMAL>;
```
should be fixed now

burumaj commented on drivers/pwm/pwm_rpi_pico.c:
```diff
@@ -0,0 +1,178 @@
+		return err;
+	}
+
+	for (int i = 0; i < NUM_PWM_SLICES; i++) {
```
done

burumaj commented on drivers/pwm/pwm_rpi_pico.c:
```diff
@@ -0,0 +1,178 @@
+		.frac = DT_INST_PROP(inst, UTIL_CAT(divider_frac_, n)),	   \
+	}
+
+#define PWM_RPI_INIT(inst)									   \
```
done

burumaj commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -135,6 +135,17 @@
+			interrupt-names = "PWM_IRQ_WRAP";
+			label = "PWM_CTRL_0";
+			status = "disabled";
+			#pwm-cells = < 3 >;
```
done

burumaj commented:
> Looking good, added a few comments. Also, make PWM a reset device. See pull requests for other drivers for reference.

@yonsch , I think I have taken all steps to pwm a reset device, but I'm not quite sure i got everything. If you could take a look at this bit in particular it would be greatly appreciated.

yonsch commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -135,6 +135,18 @@
+			clocks = <&system_clk>;
+			interrupts = <4 RPI_PICO_DEFAULT_IRQ_PRIORITY>;
+			interrupt-names = "PWM_IRQ_WRAP";
+			label = "PWM_CTRL_0";
```
Maybe `PWM_0`?

burumaj replied:
Should be updated now

yonsch reviewed (changes requested):
Thanks, only one small thing I missed last time

burumaj commented:
I rebased on main to resolve merge conflicts
